runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -435,4 +436,83 @@ function M._enable(bufnr)
   end
 end
 
+--- Returns the result IDs from the reports provided by the given client.
+--- @return lsp.PreviousResultId[]
+local function previous_result_ids(client_id)
+  local results = {}
+
+  for bufnr, state in pairs(bufstates) do
+    if state.enabled then
+      for buf_client_id, result_id in pairs(state.client_result_id) do
+        if buf_client_id == client_id then
+          table.insert(results, {
+            textDocument = util.make_text_document_params(bufnr),
+            previousResultId = result_id,
+          })
+        end
+      end
+    end
+  end
+
+  return results
+end
+
+--- @class vim.lsp.WorkspaceDiagnosticsOpts
+--- @inlinedoc
+---
+--- Only return diagnostics from the indicated client. If nil, diagnostics from all clients are returned.
+--- @field client_id? integer
+
+--- Returns workspace-wide diagnostics.
+--- @param opts? vim.lsp.WorkspaceDiagnosticsOpts
+--- @see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_dagnostics
+function M.workspace_diagnostics(opts)
```

Comment:
Before anyone starts screaming at me: Yes, I know that I said I would put this under `vim.lsp.buf` but I need to access a lot of local state in this module so I'm putting it here for now. Open to exposing the public entry point somewhere else later.